Bad CPP codegen when using var array[] argument type in imported C functions.

### Description

The following code failed to compile in cpp mode after this commit: https://github.com/nim-lang/Nim/commit/4202b606b19f032e2a34056eaa384e46b5f97496

```nim
{.emit:"""
#ifdef __cplusplus
extern "C" {
#endif

void fooArr(float data[3]) {}
void fooIntArr(int id, float data[3]) {}

#ifdef __cplusplus
}
#endif
""".}

proc fooArr(data:var array[3, cfloat]) {.importc.}
proc fooIntArr(id:cint, data:var array[3, cfloat]) {.importc, nodecl.}

#----------------------------------------------------------------
var arr = [cfloat 1,2,3]
fooArr(arr)
fooIntArr(1, arr)

```


### Nim Version

Nim Compiler Version 2.1.1 [MacOSX: arm64]
Compiled at 2024-07-05
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 4202b606b19f032e2a34056eaa384e46b5f97496
active boot switches:

### Current Output

```text
➜  cpptest git:(devel) ✗ nim cpp -f arraybug.nim                                                                                      
.................................................................../Volumes/prj/pil/test/.nimcache/@marraybug.nim.cpp:132:14: error: no matching function for call to 'fooIntArr'
        nimln_(20);     fooIntArr(((int)1), (&arr__arraybug_u6));
                        ^~~~~~~~~
/Volumes/prj/pil/test/.nimcache/@marraybug.nim.cpp:42:6: note: candidate function not viable: no known conversion from 'tyArray__JFlRG6vf7NuIEkT6kFWkgg *' (aka 'float (*)[3]') to 'float *' for 2nd argument; dereference the argument with *
void fooIntArr(int id, float data[3]) {}
     ^
1 error generated.

Error: execution of an external compiler program 'clang++ -c -std=gnu++17 -funsigned-char  -pthread -g   -I/Volumes/prj/dev/nimdev/lib -I/Volumes/prj/pil/test/cpptest -o /Volumes/prj/pil/test/.nimcache/@marraybug.nim.cpp.o /Volumes/prj/pil/test/.nimcache/@marraybug.nim.cpp' failed with exit code: 1
```


### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/ccgcalls.nim
+++ b/compiler/ccgcalls.nim
@@ -346,7 +346,8 @@ proc genArg(p: BProc, n: PNode, param: PSym; call: PNode; result: var Rope; need
     let callee = call[0]
     if callee.kind == nkSym and
         {sfImportc, sfInfixCall, sfCompilerProc} * callee.sym.flags == {sfImportc} and
-        {lfHeader, lfNoDecl} * callee.sym.loc.flags != {}:
+        {lfHeader, lfNoDecl} * callee.sym.loc.flags != {} and
+        needsIndirect:
       addAddrLoc(p.config, a, result)
     else:
       addRdLoc(a, result)

--- a/tests/ccgbugs/t23796.nim
+++ b/tests/ccgbugs/t23796.nim
@@ -1,0 +1,25 @@
+discard """
+    targets: "c cpp"
+"""
+
+# bug #23796
+
+{.emit: """
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void fooArr(float data[3]) {}
+void fooIntArr(int id, float data[3]) {}
+
+#ifdef __cplusplus
+}
+#endif
+""".}
+
+proc fooArr(data: var array[3, cfloat]) {.importc.}
+proc fooIntArr(id: cint, data: var array[3, cfloat]) {.importc, nodecl.}
+
+var arr = [cfloat 1, 2, 3]
+fooArr(arr)
+fooIntArr(1, arr)

